File: backend/src/main/java/com/echonovel/service/GenreService.java
package com.echonovel.service;

import com.echonovel.dto.request.GenreRequest;
import com.echonovel.dto.response.GenreResponse;

import java.util.List;

/**
 * Service interface for genre operations.
 */
public interface GenreService {

    /**
     * Get all genres.
     */
    List<GenreResponse> getAllGenres();

    /**
     * Create a new genre (Admin).
     */
    GenreResponse createGenre(GenreRequest request);

    /**
     * Update a genre (Admin).
     */
    GenreResponse updateGenre(Long id, GenreRequest request);

    /**
     * Delete a genre (Admin).
     */
    void deleteGenre(Long id);
}

File: frontend/src/services/genreService.ts
import api from './api';
import type { ApiResponse, GenreResponse, GenreRequest } from '../types';

export const genreService = {
  getGenres: async (): Promise<ApiResponse<GenreResponse[]>> => {
    const response = await api.get<ApiResponse<GenreResponse[]>>('/genres');
    return response.data;
  },
  
  createGenre: async (data: GenreRequest): Promise<ApiResponse<GenreResponse>> => {
    const response = await api.post<ApiResponse<GenreResponse>>('/admin/genres', data);
    return response.data;
  },

  updateGenre: async (id: number, data: GenreRequest): Promise<ApiResponse<GenreResponse>> => {
    const response = await api.put<ApiResponse<GenreResponse>>(`/admin/genres/${id}`, data);
    return response.data;
  },

  deleteGenre: async (id: number): Promise<ApiResponse<void>> => {
    const response = await api.delete<ApiResponse<void>>(`/admin/genres/${id}`);
    return response.data;
  },
};
